### ISSUE TITLE
Reduce greedy complexity to sub-quadratic?

### ISSUE THREAD

**Issue by fritzo** [dgasmith/opt_einsum#56]
The greedy strategy can be quite slow when given hundreds of tensors. Is it possible to reduce its complexity to sub-quadratic, possibly by:
- using incremental computations when possible
- using memoization when possible
- maintaining a priority queue rather than recomputing min-cost contraction at each step

**Comment by jcmgray** [comment]
It would certainly nice to speed up greedy further. Just a overview of the current algorithm:

1. Initially assess all inner contractions (i.e. those with shared indices) of ``n`` tensors. If the contraction graph has regularity ``k`` (~ average number of indices per tensor) this has cost propto ``nk`` (ignoring the cost of filtering for inner contractions which is pretty negligible). 
2. Find the minimum cost contraction (time spent doing ``min`` is currently relatively negligible also) and add that contraction to the contraction list.
3. Modify the other possible contractions that are still valid with the new result, small  cost but propto ``n`` and have to this for each of ``n`` steps.
4. Compute the approximately ``k`` new possible contractions using the new result - also for each step. With this method the same contraction should never be recomputed.

So overall the main steps of the algorithm take ``nk`` time (don't think we could do better than this), but the steps involving modifying the possible contractions take more like ``n^2`` time. Here's some profiling of a very large contraction ([generated using this gist](https://gist.github.com/jcmgray/2fb88f1b8473eef5070316f02e26d66e)):

![screenshot from 2018-08-27 23-50-44](https://user-images.githubusercontent.com/8982598/44690811-1c165300-aa54-11e8-801e-73bebb528370.png)

From that I think the biggest savings will had updating the internal format used to represent contractions. Including:

* storing the tensors in a dict (keyed by indices) rather than a list so that we don't have to manipulate index positions.
* not storing the full output set with every potential contraction (which then have to be modified). This is quite natural for small contractions, but when each of 1000 contractions is storing its 999 output index sets gets a bit inefficient.

 There might be other benefits to these kind of changes (like easier manipulation of the path once found, e.g. to ensure all constant operations are at the beginning) but it might also be quite an involved change.

**Comment by fritzo** [comment]
I've prototyped a cheaper path optimizer in https://github.com/uber/pyro/pull/1424. It is still quadratic, but is more than 10x faster on large paths. I'd like to try to further reduce its complexity, then move some of the patterns upstream.

**Comment by fritzo** [comment]
I was able to speed up [the algorithm](https://github.com/uber/pyro/pull/1424) to now 100x faster than `'greedy'` for our largest contractions. The paths are a little worse than `'greedy'`, I think because of the way I'm computing the final outer products.

My strategy was to set up data structures keyed on `input_set`, requiring the algorithm to split into four phases:
1. eagerly compute Hadamard products, thereby deduplicating keys. I think this is still optimal; we'd need to compute them anyway
2. greedily compute inner products. As @jcmgray notes, I use almost the same objective function (memory reduction, but I ignore FLOPS due to expense and complexity), so I think this is still almost optimal.
3. greedily compute outer products. I think I'm leaving something on the table here. Any ideas?
4. A final single-arg call to einsum to make sure output shape is correct. I don't really understand why this is needed, but it fixed a bunch of failing tests. Any suggestions?

Also I used static single assignment ids rather than recycled linear ids for paths, since recycled linear ids force quadratic complexity in the reductions (via Pyhon list removal), and since static single assignment ids are easier to reason about. Specifically I think it would be cleaner to `None`-out entries in an `operands` list rather than `list.pop()` them, which is expensive in the middle of a long list.

> happy to take prototypes into opt_einsum

Sure, I'll add this to the bottom of paths.py
